syft/messaging/plan.py

```diff
@@ -104,111 +95,256 @@ def append(self, key):
                 "Example: if you have self.elem1, call self.add_to_state('elem1')."
             )
         else:
-            t = getattr(self.plan, key)
-            if self.is_valid_type(t):
+            obj = getattr(self.plan, key)
+            if isinstance(obj, (FrameworkTensor, FrameworkLayerModule)):
                 self.keys.add(key)
             else:
                 raise ValueError(
-                    f"Obj of type {type(t)} is not supported in the state.\n"
+                    f"Obj of type {type(obj)} is not supported in the state.\n"
                     "Use instead tensors or parameters."
                 )
         return self
 
     def __iadd__(self, other_keys):
         return self.append(other_keys)
 
-    def is_valid_type(self, obj):
-        return isinstance(obj, (torch.nn.Module, torch.Tensor))
-
-    def get_id_at_location(self):
+    def tensors(self) -> List:
         """
-        Returns all the id_at_location of the pointers in the state
+        Fetch and return all the state elements.
+        Perform a check of consistency on the elements ids.
         """
-        id_at_location = []
-
-        for key in self.keys:
-            ptr = getattr(self.plan, key)
-            if isinstance(ptr, torch.nn.Module):
-                for param in ptr.parameters():
-                    id_at_location.append(param.id_at_location)
-            else:
-                id_at_location.append(ptr.id_at_location)
-
-        return id_at_location
+        tensors = []
+        for state_id in self.state_ids:
+            tensor = self.owner.get_obj(state_id)
+            assert tensor.id == state_id
```

Comment:
Yes usually yes, but because we use .clone(), and because of how set_() is done, we need to make sure it is still true. Actually I had a bug because this rule was violated^^
But we will remove it when the code is stable